exporter/datadogexporter/internal/metadata/internal/k8s/nodename.go

```diff
@@ -0,0 +1,63 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8sconfig"
+
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+type nodeNameProvider interface {
+	NodeName(context.Context) (string, error)
+}
+
+var _ nodeNameProvider = (*nodeNameProviderImpl)(nil)
+
+type nodeNameProviderImpl struct {
+	logger *zap.Logger
+	client k8s.Interface
+}
+
+func (p *nodeNameProviderImpl) namespace() string {
+	namespacePath := "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	val, err := ioutil.ReadFile(namespacePath)
+	if err == nil && val != nil {
+		return string(val)
+	}
+	p.logger.Warn("Could not fetch k8s namespace, using 'default'", zap.Error(err))
+	return "default"
+}
+
+func (p *nodeNameProviderImpl) NodeName(ctx context.Context) (string, error) {
+	namespace := p.namespace()
+	podName, err := os.Hostname()
```

Comment:
note: this isn't guaranteed to be the pod name

Reply:
Is there a better way to do this? I am taking the same approach as the one the Datadog Agent uses, but I trust you more than our codebase 😄

Reply:
Using the downward API would be better, but there isn't other than that.  It might be worth documenting, though, that setting the `hostname` or using `setHostnameAsFQDN` in your pod spec will break detection.